### PR TITLE
Make `jj prev` and `jj next` work when the working copy is a merge commit

### PR DESCRIPTION
**Pull request by emesterhazy**
Before this commit `jj prev` fails if the current working copy commit is a merge commit. After this commit it will prompt the user to choose the ancestor they want to select.

#2126

# Checklist

If applicable:
- [X] I have updated `CHANGELOG.md`
- [X] I have added tests to cover my changes
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)

### CODE REVIEW

File: cli/src/commands/prev.rs [hunk 1]
```diff
@@ -78,25 +77,12 @@ pub(crate) fn cmd_prev(
...
-    let ancestor_expression = RevsetExpression::commit(start_id.clone()).ancestors_at(offset);
-    let target_revset = if edit {
-        ancestor_expression
-    } else {
-        // Jujutsu will always create a new commit for prev, even where Mercurial cannot
-        // and fails. The decision and all discussion around it are available
-        // here: https://github.com/martinvonz/jj/pull/1200#discussion_r1298623933
-        //
-        // If users ever request erroring out, add `.ancestors()` to the revset below.
-        ancestor_expression.minus(&RevsetExpression::commit(current_wc_id.clone()))
```
Comment (emesterhazy) [inline cli/src/commands/prev.rs:98]: I'm not sure if we still need this comment or not, or where would be a better place to put it. Maybe it belongs above the call to `tx.check_out` further down?

**Comment by emesterhazy** [comment]
@PhilipMetzger I think that this just works, and that the logic is actually a bit simpler now. You are probably the expert for this command though. Do you think there are any other scenarios we should test?

File: cli/src/commands/prev.rs (hunk 1 above)
Comment (PhilipMetzger) [reply cli/src/commands/prev.rs:98]: Ok, after looking at https://github.com/martinvonz/jj/pull/1200#discussion_r1253946006, where the check originally comes from, it'd be better to prompt here instead of just skipping the commit and dropping the optional check.

Comment (PhilipMetzger) [reply cli/src/commands/prev.rs:98]: cc @martinvonz, do you remember the intention of it?

Comment (emesterhazy) [reply cli/src/commands/prev.rs:98]: I guess I don't really understand the comment in the first place. In what scenario might we want to return an error?

Comment (martinvonz) [reply cli/src/commands/prev.rs:98]: No, I don't know what that comment was about either.

**Comment by martinvonz** [comment]
It looks like `jj next` could use the same fix. Feel like taking care of that too?

**Comment by emesterhazy** [comment]
Yeah I can take a look at `next` next. Should it go into this PR as a separate commit, or should we merge this first? I'm fine either way.

**Comment by martinvonz** [comment]
> Yeah I can take a look at `next` next. Should it go into this PR as a separate commit, or should we merge this first? I'm fine either way.

I would do it in the same PR but up to you since you're the one doing it.

File: cli/src/commands/prev.rs [hunk 2]
```diff
@@ -78,25 +76,12 @@ pub(crate) fn cmd_prev(
...
-    };
+    // This is the distance of the target ancestor from the current working copy
+    // revision (@). The offset provided by the user is the distance of the
+    // target ancestor from the parent(s) of the working copy revision (@-).
+    let target_ancestor_distance = if edit { args.offset } else { args.offset + 1 };
```
Comment (yuja) [inline cli/src/commands/prev.rs:82]: nit: to make it look similar to `cmd_next()`, it might be better to stick to the original structure and turn `start_id` into `Vec<CommitId>` or `RevsetExpression` instead.

File: cli/src/commands/prev.rs (hunk 1 above)
Comment (PhilipMetzger) [reply cli/src/commands/prev.rs:98]: I wasn't interested in the comment, but why we'd use the parent ids in the revset. It was your suggestion back then.

File: cli/src/commands/prev.rs (hunk 2 above)
Comment (emesterhazy) [reply cli/src/commands/prev.rs]: I pushed a new revision using revset expressions instead of arithmetic on the offset. Please take a look :)

File: cli/src/commands/prev.rs [hunk 3]
```diff
@@ -79,24 +77,15 @@ pub(crate) fn cmd_prev(
...
-            _ => return Err(user_error("Cannot run `jj prev` on a merge commit")),
-        }
-    };
-    let ancestor_expression = RevsetExpression::commit(start_id.clone()).ancestors_at(offset);
+    // If --edit is specified or implied, then the offset is relative to the
+    // current working copy commit. Otherwise, the offset is relative to the
+    // parent(s) of the current wc commit.
     let target_revset = if edit {
-        ancestor_expression
+        RevsetExpression::commit(current_wc_id.clone()).ancestors_at(args.offset)
```
Comment (emesterhazy) [inline cli/src/commands/prev.rs:84]: The documentation for `ancestors_at` seems wrong:

```
    /// Ancestors of `self`, including `self` until `generation` back.
    pub fn ancestors_at(self: &Rc<RevsetExpression>, generation: u64) -> Rc<RevsetExpression> {
        self.ancestors_range(generation..(generation + 1))
    }
 ```
Based on the docstring the `target_revset` should include the wc commit and all of the ancestors back to the specified offset. Instead it looks like `ancestors_at` actually returns only the ancestors at the specified generation number.

I think this is either a bug in the documentation or the implementation. Perhaps the docs are wrong?

@yuja

**Comment by emesterhazy** [comment]
I pushed a second commit that removes the merge commit limitation from `jj next` as well.

File: cli/src/commands/prev.rs [hunk 4]
```diff
@@ -78,25 +76,15 @@ pub(crate) fn cmd_prev(
...
-    } else {
-        match current_wc.parent_ids() {
-            [parent_id] => parent_id,
-            _ => return Err(user_error("Cannot run `jj prev` on a merge commit")),
-        }
-    };
-    let ancestor_expression = RevsetExpression::commit(start_id.clone()).ancestors_at(offset);
+    // If --edit is specified or implied, then the offset is relative to the
+    // current working copy commit. Otherwise, the offset is relative to the
+    // parent(s) of the current wc commit.
```
Comment (yuja) [inline cli/src/commands/prev.rs:81]: nit: I think the comment in `cmd_next()` is clearer.
```
    // If we're editing, start at the working-copy commit. Otherwise, start from
    // its direct parent(s).
```

File: cli/src/commands/prev.rs [hunk 5]
```diff
@@ -107,14 +95,16 @@ pub(crate) fn cmd_prev(
         [target] => target,
         [] => {
             return Err(user_error(format!(
-                "No ancestor found {offset} commit{} back",
-                if offset > 1 { "s" } else { "" }
+                "No ancestor found {} commit{} back",
+                args.offset,
```
Comment (yuja) [inline cli/src/commands/prev.rs:99]: nit: either revert this change or update `cmd_next()` to use `args.offset`

File: cli/tests/test_next_prev_commands.rs [hunk 6]
```diff
@@ -248,9 +248,87 @@ fn test_prev_fails_on_merge_commit() {
     "###);
 
     // Try to advance the working copy commit.
```
Comment (yuja) [inline cli/tests/test_next_prev_commands.rs:356]: nit: "advance" sounds odd. maybe we can remove this comment since `prev` works now.

File: cli/src/commands/next.rs [hunk 7]
```diff
@@ -115,24 +115,18 @@ pub(crate) fn cmd_next(
...
-    };
-    let descendant_expression = RevsetExpression::commit(start_id.clone()).descendants_at(offset);
-    let target_expression = if edit {
-        descendant_expression
-    } else {
-        descendant_expression.minus(&RevsetExpression::commit(current_wc_id.clone()).descendants())
+        wc_revset
+            .parents()
+            .descendants_at(args.offset)
+            .minus(&wc_revset)
```
Comment (yuja) [inline cli/src/commands/next.rs:127]: Did you remove `wc_revset.descendants()` by intention?

I think `.descendants()` was redundant because `--edit` is implied now, but it's worth mentioning.

File: cli/src/commands/prev.rs (hunk 3 above)
Comment (yuja) [reply cli/src/commands/prev.rs:83]: Yes, the doc is wrong. As the function name suggests, it should return `generation`-th ancestors counted from the `self`.

**Review by yuja — APPROVED** [review]
lgtm, thanks.

File: cli/src/commands/prev.rs (hunk 4 above)
Comment (emesterhazy) [reply cli/src/commands/prev.rs:81]: Updated.

File: cli/tests/test_next_prev_commands.rs (hunk 6 above)
Comment (emesterhazy) [reply cli/tests/test_next_prev_commands.rs:356]: Yeah, "advance" doesn't sound right. I think it might have just been copied over from the tests for `jj next` when these were written. I deleted this comment and the identical one in the other `jj prev` merge test.

File: cli/src/commands/prev.rs (hunk 5 above)
Comment (emesterhazy) [reply cli/src/commands/prev.rs:99]: Thanks. I decided to delete the offset variable in next.rs as well. There isn't a good reason for it to exist and I think it's clearer if we just access the offset through the args struct.

File: cli/src/commands/next.rs (hunk 7 above)
Comment (emesterhazy) [reply cli/src/commands/next.rs:127]: Can you clarify the question? I'm not sure I understand what you're asking about.

If you're asking about the `.minus`, that's intentional.

File: cli/src/commands/prev.rs (hunk 3 above)
Comment (emesterhazy) [reply cli/src/commands/prev.rs:82]: Sounds good. Suggesting an edit in https://github.com/martinvonz/jj/pull/3471

File: cli/src/commands/next.rs (hunk 7 above)
Comment (yuja) [reply cli/src/commands/next.rs:126]: The query is changed from `.minus(&wc_revset.descendants())` to `.minus(&wc_revset)`.

The original query was more correct because `@::` shouldn't be considered the "next" candidates, but that wouldn't matter anymore since `--edit` is implied if `@` isn't a head. The new query is optimal.

While I think the new query is better, this is subtle change, and should be documented in commit description or comment.

**Review by PhilipMetzger — APPROVED** [review]
LG

File: cli/src/commands/next.rs (hunk 7 above)
Comment (emesterhazy) [reply cli/src/commands/next.rs:126]: > but that wouldn't matter **anymore** since --edit is implied if @ isn't a head.

Ah. I wasn't around when this command was originally implemented and I thought it always worked this way. Looking through the commit history it looks like this is actually a pretty recent change from https://github.com/martinvonz/jj/pull/3028.

For what it's worth I'm not sure if the old behavior for `next` would be very ergonomic since you could be in the middle of a linear history and `jj next` would refuse to move.

I added a comment above the call to `.minus()`.

**Comment by emesterhazy** [comment]
Thanks everyone for helping to review this!
